cpu/esp32/esp-now/esp_now_netif.c

```diff
@@ -0,0 +1,250 @@
+    }
+
+    /* pkt has been copied into payload_iolist, we're done with it. */
+    gnrc_pktbuf_release(pkt);
+
+    DEBUG("gnrc_esp_now: sending packet to %02x:%02x:%02x:%02x:%02x:%02x with size %u\n",
+          mac[0], mac[1], mac[2], mac[3], mac[4], mac[5], (unsigned)payload_iolist.iol_len);
+
+    return dev->driver->send(dev, &iolist);
+}
+
+static gnrc_pktsnip_t *_recv(gnrc_netif_t *netif)
+{
+    netdev_t *dev = netif->dev;
+    esp_now_netdev_t *esp_now = (esp_now_netdev_t*)dev;
+
+    int bytes_expected = dev->driver->recv(dev, NULL, 0, NULL);
+    if (bytes_expected <= 0) {
+        DEBUG("gnrc_esp_now: failed receiving packet: %d\n", bytes_expected);
+        return NULL;
+    }
+
+    gnrc_pktsnip_t *pkt;
+    pkt = gnrc_pktbuf_add(NULL, NULL,
+                          bytes_expected,
+                          GNRC_NETTYPE_UNDEF);
+    if (!pkt) {
+        DEBUG("gnrc_esp_now: cannot allocate pktsnip.\n");
+
+        /* drop the packet */
+        dev->driver->recv(dev, NULL, bytes_expected, NULL);
+
+        return NULL;
+    }
+
+    int nread = dev->driver->recv(dev, pkt->data, bytes_expected, NULL);
+    if (nread <= 0) {
+        DEBUG("gnrc_esp_now: read error %d\n", nread);
+        goto err;
+    }
+
+    if (nread < bytes_expected) {
+        DEBUG("gnrc_esp_now: reallocating.\n");
+        gnrc_pktbuf_realloc_data(pkt, nread);
+    }
+
+    gnrc_pktsnip_t *mac_hdr;
+    mac_hdr = gnrc_pktbuf_mark(pkt, ESP_NOW_ADDR_LEN, GNRC_NETTYPE_UNDEF);
+    if (!mac_hdr) {
+        DEBUG("gnrc_esp_now: no space left in packet buffer\n");
+        goto err;
+    }
+
+    gnrc_pktsnip_t *esp_hdr;
+    esp_hdr = gnrc_pktbuf_mark(pkt, sizeof(esp_now_pkt_hdr_t), GNRC_NETTYPE_UNDEF);
+    if (!esp_hdr) {
+        DEBUG("gnrc_esp_now: no space left in packet buffer\n");
+        pkt = mac_hdr;
+        goto err;
+    }
+    esp_now_pkt_hdr_t *hdr = (esp_now_pkt_hdr_t*)esp_hdr->data;
+
+#ifdef MODULE_L2FILTER
+    if (!l2filter_pass(dev->filter, mac_hdr->data, ESP_NOW_ADDR_LEN)) {
+        DEBUG("gnrc_esp_now: incoming packet filtered by l2filter\n");
+        pkt = mac_hdr;
+        goto err;
+    }
+#endif
+
+    switch (hdr->flags) {
+#ifdef MODULE_GNRC_SIXLOWPAN
+        case 1:
+            pkt->type = GNRC_NETTYPE_SIXLOWPAN;
+            break;
+#endif
+        default:
+            pkt->type = GNRC_NETTYPE_UNDEF;
+    }
+
+    gnrc_pktsnip_t *netif_hdr;
+    netif_hdr = gnrc_pktbuf_add(
+                    NULL,
+                    NULL,
+                    sizeof(gnrc_netif_hdr_t) + 2 * ESP_NOW_ADDR_LEN,
+                    GNRC_NETTYPE_NETIF);
+    if (!netif_hdr) {
+        DEBUG("gnrc_esp_now: no space left in packet buffer\n");
+        pkt = mac_hdr;
+        goto err;
+    }
+
+    gnrc_netif_hdr_init(netif_hdr->data, ESP_NOW_ADDR_LEN, ESP_NOW_ADDR_LEN);
+    gnrc_netif_hdr_set_src_addr(netif_hdr->data, mac_hdr->data, ESP_NOW_ADDR_LEN);
+    gnrc_netif_hdr_set_dst_addr(netif_hdr->data, esp_now->addr, ESP_NOW_ADDR_LEN);
```

Comment:
Use

```C
gnrc_netif_hdr_t *netif_hdr = gnrc_netif_hdr_build(
        mac_hdr->data, ESP_NOW_ADDR_LEN,
        esp_now->addr, ESP_NOW_ADDR_LEN,
    )
```

instead of all of the above. Makes it easier to read and saves some code duplication ;-).

Reply:
See also #10661 for some enhancement for the line below ;-)

Reply:
I'd like to avoid a dependency on that for now, so I just changed the line to use netif->pid for now. Can be changed here once #10661 is merged, or it itself can just change it here if this gets merged first.
gnrc_netif_hdr_build is used like suggested now.

Reply:
Yeah, I just added the #10661 note in case it is merged first ;-).